### PR TITLE
WIP: Semi supervised

### PR DESCRIPTION
**Pull request by ThierryJudge**
## Summary:
Work in progress: Add support for semi-supervised learning for Pytorch-Lightning

### Features:
* Added SemiSupervisedIterator (
* Added SSLModule

### TODO
* Test Pi model

## Checklist:

* [ ] Your code is documented (To validate this, add your module to `tests/documentation_test.py`).
* [ ] Your code is tested with unit tests.
* [ ] You moved your Issue to the PR state.

### CODE REVIEW

File: src/baal/utils/pytorch_lightning.py [hunk 1]
```diff
@@ -70,3 +98,38 @@ def predict_on_dataset_generator(self, *args, **kwargs):
...
+
+        Args:
+            pool (iterable): dataset pool indices.
+
+        Returns:
+            boolean, Flag indicating if we continue training.
+
+        """
+        # High to low
+        if pool is None:
```
Comment (Dref360) [inline src/baal/utils/pytorch_lightning.py:114]: In this case we can remove everything and just call self.predict_on_dataset_generator(**self.kwargs) I think? We will call the pool_loader.

File: src/baal/utils/ssl_iterator.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+        if dl_2 is not None:  # Avoid error if p=1
+            self.dl_2 = dl_2
+            self.dl_2_iter = cycle(dl_2)
+            self.len_dl2 = len(dl_2)
+        else:
+            p = 1
+
+        self.num_steps = num_steps or (self.len_dl1 + self.len_dl2)
+        self.p = None if p is None else [p, 1 - p]
+        self._pool = None
```
Comment (Dref360) [inline src/baal/utils/ssl_iterator.py:35]: unused I think.

File: src/baal/utils/pytorch_lightning.py [hunk 3]
```diff
@@ -45,6 +48,31 @@ def on_train_start(self, trainer, module):
...
+        heuristic (Heuristic): Heuristic from baal.active.heuristics.
+        ndata_to_label (int): Number of sample to label per step.
+        max_sample (int): Limit the number of sample used (-1 is no limit).
+        **kwargs: Parameters forwarded to `get_probabilities` and to pytorch_ligthning Trainer.__init__
+    """
+
+    def __init__(self, dataset: ActiveLearningDataset,
+                 heuristic: heuristics.AbstractHeuristic = heuristics.Random(),
+                 ndata_to_label: int = 1,
+                 max_sample=-1,
```
Comment (Dref360) [inline src/baal/utils/pytorch_lightning.py:66]: Will need to be removed too.

File: src/baal/utils/pytorch_lightning.py [hunk 4]
```diff
@@ -45,6 +48,31 @@ def on_train_start(self, trainer, module):
 
 
 class BaalTrainer(Trainer):
+    """Object that perform the training and active learning iteration.
+
+    Args:
+        dataset (ActiveLearningDataset): Dataset with some sample already labelled.
+        get_probabilities (Function): Dataset -> **kwargs ->
```
Comment (Dref360) [inline src/baal/utils/pytorch_lightning.py:55]: unused I think?

**Review by Dref360 — COMMENTED** [review]
Will need to add tests for BaaLTrainer.step.

**Comment by Dref360** [comment]
Coudl also move the ssl experiment files in their own folder. Ex `experiments/ssl_experiments`

File: experiments/pimodel_mcdropout_cifar10.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    print("Active set length: {}".format(len(active_set)))
+    print("Pool set length: {}".format(len(active_set.pool)))
+
+    heuristic = BALD()
+    net = vgg16(pretrained=False, num_classes=10)
+    model = PIActiveLearningModel(network=net, active_dataset=active_set, hparams=params)
+
+    dp = 'dp' if params.gpus > 1 else None
+    trainer = BaalTrainer(max_epochs=params.epochs, default_root_dir=params.data_root,
+                          gpus=params.n_gpus, distributed_backend=dp,
```
Comment (Dref360) [inline experiments/pimodel_mcdropout_cifar10.py:78]:
```suggestion
                          gpus=params.gpus, distributed_backend=dp,
```

File: experiments/pimodel_mcdropout_cifar10.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    active_set = ActiveLearningDataset(
+        CIFAR10(params.data_root, train=True, transform=PIModel.train_transform, download=True),
+        pool_specifics={'transform': PIModel.test_transform},
+        make_unlabelled=lambda x: x[0])
+    active_set.label_randomly(100)
+
+    print("Active set length: {}".format(len(active_set)))
+    print("Pool set length: {}".format(len(active_set.pool)))
+
+    heuristic = BALD()
```
Comment (Dref360) [inline experiments/pimodel_mcdropout_cifar10.py:72]: This should be configurable to compare Random with BALD or Entropy

**Comment by Dref360** [comment]
Alright, I think we are ready for review @parmidaatg :)

File: experiments/pl_baal_example.py [hunk 7]
```diff
@@ -162,16 +161,19 @@ def main(hparams):
...
-                              ndata_to_label=hparams.query_size)
+                          callbacks=[ResetCallback(copy.deepcopy(model.state_dict()))],
+                          dataset=active_set,
+                          heuristic=heuristic,
+                          ndata_to_label=hparams.query_size
+                          )
 
     AL_STEPS = 100
     for al_step in range(AL_STEPS):
+        # TODO fix this
```
Comment (Dref360) [inline experiments/pl_baal_example.py:172]:
```suggestion
        # TODO Issue 95 Make PL trainer epoch self-aware
```

**Review by parmidaatg — APPROVED** [review]

**Review by rafapi — APPROVED** [review]
LGTM

**Pull request merged**